[1단계 방탈출 결제 / 배포] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
안녕하세요 매트 리뷰 감사합니다. :)

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [부기의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4408)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
크게 객체지향을 고려하지 않음

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?

### 학습한 내용

##### 여러 HTTP client 학습
LMS에 작성된 OkHttp, Retrofit, RestClient, WebClient, RestTemplate를 각각 학습해보았습니다.
- OkHttp는 자바 진영에서 사용하는 경량 HTTP client 라이브러리로 저수준 라이브러리입니다.
- Retrofit은 자바 진영에서 사용하는 HTTP client로 OkHttp를 사용하여 좀 더 고수준에서 쉽게 사용할 수 있도록 만든 라이브러리입니다.
	- OkHttp와 Retrofit은 스프링 보다는 안드로이드 진영에서 사용하는 라이브러리로, 스프링 환경을 고려하지 않은 외부 라이브러리이기 때문에 비교적 환경 적합성이 떨어질 수 있다고 판단하였습니다.
- RestClient는 스프링에서 제공하는 Http Client로 Fluent 패턴으로 작성되어 선언형으로 사용할 수 있는 장점이 있었습니다. 
- RestTemplate는 스프링에서 제공하는 Http Client로 메서드 체이닝보다는, 메서드 호출을 통해서 사용합니다.
- WebClient는 비동기로 Http 요청을 보낼 수 있는 메서드로 아직 WebFlux에 대해서 학습하지 못해 사용하지는 않았습니다.!

그래서 이 중에서 스프링 환경에 적합하고, Fluent 패턴으로 작성되어 선언형으로 사용할 수 있는 RestClient를 선정해서 사용하였습니다.
- 확실히 메서드 체이닝으로 무엇을 하는지 작성하니 RestTemplate로 작성하는 것보다 코드의 가독성이 높다고 생각합니다.

##### RestClient
RestClient는 스프링에서 빈으로 등록해주지 않지만, RestClient.Builder를 빈으로 등록해줍니다.
아마 가추법으로 생각해보면 스프링에서 Builder를 주입받아서 각 클래스마다 사용하는 옵션들을 각자 설정해서 사용하라는 의미가 아닐까 생각해보았습니다.

##### @RestClientTest
RestClient를 사용하는 클래스를 테스트하기 위해 필요한 클래스들만 빈으로 로드해주는 어노테이션입니다.
- value에 테스트하려는 클래스를 담아서 해당 클래스를 빈으로 등록하여 줍니다.

`MockRestServiceServer`를 사용해서 내가 요청을 보내려는 서버에 대해서 행위 검증과 Stubbing을 해줄 수 있습니다.
- 이 클래스는 RestClient 내부에 있는 http 요청을 만들어 전송해주는 `ClientHttpRequestFactory` 구현체를 Mock으로 대체해서 우리가 테스트하려는 클래스가 외부로 전송하는 요청을 이 클래스를 통해 잡아서 검증하고, Stubbing 해줍니다.

우리가 테스트하려는 클래스가 스프링이 등록한 RestClient.Builder를 주입받아서 구현해야만, 이 `MockRestServiceServer`가 이 RestClient 내부를 모킹해줄 수 있습니다.



##### MockServerRestClientCustomizer
만약, RestClient 빈을 여러개 등록해서 사용중이라면, MockServer가 여러개 생길 수 있는데, 이 때 MockServerRestClientCustomizer를 사용하면, Key-Value 형태의 MockServer를 저장하고 있습니다.
따라서, RestClient에 맞는 MockServer를 MockServerRestClientCustomizer를 통해 알맞게 받아 사용할 수 있습니다.

##### 제네릭 타입 소거
RestClient 학습 테스트 도중 ParameterizedTypeReference<>() {}를 익명 클래스로 사용해서 사용했습니다.
- 사용한 이유는 자바 컴파일러가 런타임 시점에 필요없는 제네릭 타입 정보들을 모두 제거하는 과정을 거치는데, 익명 클래스의 경우 제네릭 타입이 소거가 되지 않기 때문에 컴파일러를 회피하여 제네릭 정보를 저장하도록 만들기 위해서 였습니다.

RestClient에게 List<Type정보>를 담아서 List 타입 정보 뿐만 아니라 List의 제네릭 타입 정보까지 제공해줄 수 있었습니다.

##### DTO 역직렬화
PaymentRequest와 ReservationRequest를 분리하여 하나의 DTO인, ReservationPaymentRequest를 만들어주었습니다. 
하지만, 실제 DTO는 json상에서 depth를 가지도록 하고 싶지 않았기 때문에  Jackson 라이브러리에서 제공해주는 `@JsonUnwrapped`를 사용하여 평탄화 해주었습니다.
하지만, record로 이 클래스를 정의해주었더니, 다음의 문제가 있었습니다.
1. jackson은 `@JsonUnwrapped` 어노테이션을 통해 평탄화 작업을 해줄 때, 각 내부 객체들을 하나씩 생성해서 단계적으로 객체를 생성해줍니다. 
2. 하지만 record의 경우 필드가 final로 정의되어 있기 때문에 하나씩 생성해줄 수 없었고 예외가 발생하였습니다.

그래서 이를 해결해주기 위해 lombok의 `@Builder`, `@Jacksonized`를 통해 이를 해결해주었습니다.
- `@Builder`는 해당 클래스에 빌더 패턴을 통해 생성할 수 있는 빌더 메서드를 만들어줍니다.
- `@Jacksonized`는 해당 클래스에 생성 방식을 jackson에게 builder 메서드를 통해 수행하라고 알려줍니다.

이를 통해 jackson 라이브러리가 record 객체도 단계적으로 생성할 수 있도록 만들어 역직렬화시 예외가 발생하지 않았습니다.


##### timeout 설정 이슈

`@RestClientTest`로 테스트를 하던 도중, `RestClient`가 타임 아웃을 설정해주지 않으면 디폴트 값으로 무제한 대기를 한다는 사실을 알았습니다.
그래서 심각한 문제라고 생각하고, 이를 적용해주었습니다.
다만, 설정하면서 문제가 있었습니다.

timeout을 설정하기 위해 앞에서 말한 RestClient의 `ClientHttpRequestFactory`를 클래스 내부 생성자에서 직접 생성해서 .requestFactory()를 통해 설정해주어야 합니다.
그래서 이렇게 해주었더니 `@RestClientTest`에서 Mock으로 치환했던 `ClientHttpRequestFactory`
객체가 생성자에서 다시 한번 치환되는 일이 발생했습니다.

그래서 Mock이 아닌 실제 토스 서버로 요청이 가버려 테스트가 깨져버렸습니다.

이를 해결해주기 위해 찾은 방법은 다음의 issue에서 찾았습니다.
- https://github.com/spring-projects/spring-boot/issues/38832 

이 issue에서는 2가지 방법을 제시하고 있었습니다.
```java
  
  @Bean
  public RestClientCustomizer restClientCustomizer() {
    return (restClientBuilder) -> restClientBuilder
            .requestFactory(new JdkClientHttpRequestFactory())
            .baseUrl(baseUrl);
  }
```
- RestClientCustomizer를 빈으로 등록해 전역 RestClient 옵션을 지정해줄 수 있었습니다.

```yml
spring:
	http:  
	  client:  
		read-timeout: 30000  
		connect-timeout: 3000
```
- application.yml에 위와 같이 설정해서 전역 옵션으로 지정해줄 수 있었습니다.

굳이 빈으로 등록해줄 필요 없다고 판단하였고 아직 외부 api 요청이 하나밖에 없어 일단 전역 설정을 통해 지정해주었습니다.



### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.


### 리뷰 범위 

아래에 이번 미션에서 변경된 부분인 커밋 내역만 포함된 file changed 입니다. ! 리뷰 감사합니다. 😃매트😇 
[리뷰 범위 링크 바로가기](https://github.com/woowacourse/spring-roomescape-payment/pull/190/files/a743ca8ff23bfe372ed6ef2e7f13e294ae897237..4bf57d8d6775c1ba2a8551d63767283eb5f85bad)


### 결제 로직 흐름

```mermaid
graph TD
    A[트랜잭션 시작] --> B[예약 검증]
    B --> C[예약 생성]
    C --> D{예약 생성 성공 여부}
    D -- 실패 --> E[예외 처리 후 종료 - 트랜잭션 롤백]
    D -- 성공 --> F[결제 승인 API 호출]
    F --> G{결제 승인 성공 여부}
    G -- 성공 --> H[정상 처리 및 트랜잭션 커밋]
    G -- 실패 --> I[예외 발생 후 트랜잭션 자동 롤백]
```

Hacanna42 committed:
chore: 이전 미션 코드 가져오기

Hacanna42 committed:
chore: order ID prefix를 변경

Hacanna42 committed:
docs: todo 체크

Hacanna42 committed:
feat: reservation request에 결제 정보 추가

Hacanna42 committed:
fix: 예약 대기 요청이 Payment 정보를 갖지 않도록 수정

Hacanna42 committed:
feat: 예약 시 토스 결제 연동

- 예약 생성 시 트랜잭션으로 범위를 묶음

Hacanna42 committed:
test: 토스 결제 예약 관련 테스트 추가

Hacanna42 committed:
refactor: ObjectMapper 필드 분리

Hacanna42 committed:
docs: 프로젝트 설명 문서 추가

changuii committed:
refactor: key 추출

changuii committed:
test: PaymentManager 테스트 코드 추가 및 toss-payment-key 환경 변수 적용

- PaymentManager의 `confirmPayment` 메서드에 대한 유닛 테스트 코드 추가
- toss API 응답 코드(200, 400, 500)에 따른 예외 처리 테스트 작성
- 기존 `secret.payment-key`를 `secret.toss-payment-key`로 명칭 변경
- PaymentManager에 `@RequiredArgsConstructor` 추가 및 DI 방식 수정

changuii committed:
test: MockRestServiceServer 활용으로 테스트 리팩토링 및 신규 테스트 케이스 추가

RestClient.Builder 대신 MockRestServiceServer를 사용하도록 변경하여 테스트 구조를 단순화하였으며,
서버 지연 상황을 검증하는 테스트 케이스를 새로 추가함.

changuii committed:
feat: toss 결제 예외 처리 추가 및 테스트 수정

- toss API 호출 시 타임아웃(ResourceAccessException) 발생 처리 로직 추가
- PaymentError에 연결 실패 에러 메시지 반환 메서드(connectFail) 추가
- PaymentManagerTest에 타임아웃 예외 테스트 케이스 추가
- application.yml에 http client timeout 설정 추가 (readTimeout, connectTimeout)

changuii committed:
refactor: ServerInternalException 클래스 상속 변경 및 HttpStatus 추가

changuii committed:
refactor: JwtProvider 생성자 주입으로 변경

JwtProvider의 필드를 생성자 주입 방식으로 수정하여 코드 가독성과 테스트 용이성을 향상시킴. `@PostConstruct` 제거.

changuii commented on src/main/java/roomescape/auth/JwtProvider.java:
```diff
@@ -15,15 +14,14 @@
+    ){
+        this.validityInMilliseconds = validityInMilliseconds;
         this.key = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKey));
     }
```
불변으로 안전하게 보관하기 위해, `final` 키워드를 추가하고 생성자에서 초기화해주도록 변경하였습니다.

hyeonic commented on src/main/java/roomescape/auth/dto/LoginResponse.java:
```diff
@@ -0,0 +1,12 @@
+
+public record LoginResponse(String name) {
+
+    public LoginResponse{
+        if(name == null){
+            throw new ResponseInvalidException();
+        }
+    }
```
컨벤션 확인 한번 부탁드려요!

changuii replied:
감사합니다!

hyeonic commented on src/main/java/roomescape/common/PaymentManager.java:
```diff
@@ -0,0 +1,57 @@
+
+// toss pg사 이용
+@Component
+@RequiredArgsConstructor
```
`@RequiredArgsConstructor`의 역할은 무엇인가요? 아래 생성자 주입과 중복된 역할을 하진 않을까요?

changuii replied:
맞습니다. 중복된 역할을 하고있어서 제거하였습니다. :)

hyeonic commented on src/main/java/roomescape/reservation/dto/ReservationPaymentRequest.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.extern.jackson.Jacksonized;
+import roomescape.common.dto.PaymentRequest;
+
+@Builder
+@Jacksonized
+public record ReservationPaymentRequest(
+        @JsonUnwrapped @Valid ReservationRequest reservationRequest,
+        @JsonUnwrapped @Valid PaymentRequest paymentRequest
+) {
+}
```
`@JsonUnwrapped`이런 애노테이션이 있었군요! 관련하여 단점은 없을까요?

changuii replied:
단점에 대해서 찾아보았습니다. 단점은 생각보다 이렇게 json을 평탄화하는 작업이 복잡하기 때문에 이렇게 하나의 depth가 아니라 여러 depth가 발생하거나, 복잡한 구조의 DTO를 평탄화한다면 버그가 많이 발생한다고 합니다.
- 또한, 상위에 있는 필드 명이 겹치는 문제도 고려해주어야 합니다.

그리고 앞서 말했듯이 `@JsonUnwrapped`와 자바 record를 사용할 때 문제점이 있고 이를 lombok을 통해 우회적으로 동작하도록 구현해두었기 때문에 추후에 문제가 발생할 가능성이 존재한다고 생각합니다.

따라서 왠만하면 이렇게 평탄화하기 보다는 그냥 프론트와 협업을 통해 depth를 추가하여 js에서 데이터 depth를 들어가서 추출하거나 dto를 재사용하지 않도록 구현할 것 같습니다.

hyeonic replied:
옙 좋습니다 👍

hyeonic commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,209 @@
+    @Transactional
+    public ReservationResponse create(final ReservationPaymentRequest request, final LoginMember loginMember) {
+        final ReservationTime reservationTime = getReservationTimeById(request.reservationRequest().timeId());
+        final Theme theme = getThemeById(request.reservationRequest().timeId());
```
관련 내용 같이 확인해주세요~

hyeonic commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,209 @@
+        final Theme theme = getThemeById(request.reservationRequest().timeId());
+        final Member member = getMemberByEmail(loginMember.email());
+
+        final LocalDateTime currentTimestamp = LocalDateTime.now();
```
우리가 제어할 수 있는 영역이 존재하면 테스트 시 어려울 수 있을 것 같아요. 어떻게 개선할 수 있을까요?

changuii replied:
맞습니다. 제어할 수 없는 LocalTime.now()에 의존하고 있어 이를 테스트에서 제어해주기가 어려웠습니다.
이를 해결해주기 위해 Service에서 직접 사용하기 보다는, 제어할 수 있는 클래스를 의존해주고 테스트에서 쉽게 stubbing을 할 수 있도록 변경해주었습니다.

hyeonic commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -0,0 +1,88 @@
+@Entity
+@Getter
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
id로만 equals & hashCode를 재정의한 이유가 있을까요??

changuii replied:
저는 Entity가 데이터베이스 테이블에서 하나의 row를 표현하는 객체라고 생각했습니다. 데이터베이스에서는 기본 키를 통해 하나의 row를 식별할 수 있고, entity 객체도 id를 통해 동등성을 비교하면 충분하다고 판단하였습니다.

hyeonic replied:
JPA의 Entity는 LAZY 등과 같은 기능 제공을 위해 프록시를 적극 활용하고 있는데요, 단순한 id 비교를 통해서만 진행할 때 실제 객체와 프록시 객체 사이의 비교와 같은 케이스도 동일하게 보장할 수 있을까요??

changuii replied:
매트가 말씀하신대로,, LAZY를 통해 존재하는 프록시 객체도 이렇게 단순한 id 비교를 통해 동등성 검사가 가능할까를 테스트해보았습니다.

환경은 customer가 orders를 하나 가지는 다대일 관계로 매핑해주었습니다.
이때, customer를 조회할 때 orders를 LAZY로 가져오도록 하였고, Orders는 id를 기준으로 EqualsAndHashCode를 구현하였습니다.

### 1. 테스트 성공

```java
@DisplayName("equals and hashCode test")  
@Test  
void test() {  
    // 1.비교할 대상인 orders 생성  
    final Orders orders = new Orders(1L);  
  
    // 2.실제로 데이터베이스에 저장할 (프록시로 조회할) orders 저장  
    final Orders o = new Orders();  
  
    // 3. orders와 다대일 관계를 가지는 customer 생성 (orders를 연관관계로 가짐)  
    final Customer customer = new Customer("first", "last", o);  
  
    // 4. orders와 customer를 persist하고 flush를 통해 데이터베이스에 저장  
    entityManager.persist(o);  
    entityManager.persist(customer);  
    entityManager.flush();  
    entityManager.clear();  
  
    // 5. entity manager로 다시 조회  
    final Customer savedCustomer = entityManager.find(Customer.class, 1L);  
  
  
    // 6. proxy 객체와 1번에서 만든 비교 대상을 equals()로 비교  
    assertThat(savedCustomer.getOrders().equals(orders)).isTrue();  
}
```
- 위 코드처럼 savedCustomer.getOrders().equals(orders)를 호출한 경우에는 기대한 true를 반환하였습니다.

### 2. 테스트 실패
하지만, 반대로 orders.equals(savedCustomer.getOrders())를 호출한 경우 기대하지 않은 False가 반환되었습니다.
찾아보니, equals()를 구현할 때 사용하는 getClass()에서 프록시 타입인 경우 proxy class 값과 비교되기 때문에 equals()가 false로 반환되고 있다는 것을 알았습니다.
- Lombok 제거 후 intellij로 생성하여 테스트

### Hibernate.getClass() 도입
hibernate에서 프록시 객체의 실제 getClass()로 가져오는 메서드를 제공하고 있어 이를 도입하였습니다.
- 반대로 프록시 객체를 넣지 않으면 문제가 생기지 않을까 궁금했는데, 프록시 객체가 아니라면 그냥 getClass()를 반환해주고 있었습니다.

그래도 테스트가 실패했습니다.
byte buddy로 만든 프록시 객체가 프록시로 사용하기 위해 서브 클래스로 생성되어 있어 id필드가 존재했고 이 id필드에는 null이 담겨있어서 필드로 바로 접근했을 경우 프록시 객체의 id가 조회되어 null과 비교하였고 결국 false가 반환되었습니다.

![Pasted image 20250601233152](https://github.com/user-attachments/assets/a6d6d55c-9328-451e-b13f-c677e4a83479)

### 3. getId()로 변경
그래서 지연 로딩을 통해 getId()를 조회하도록 만들어 주었더니 성공했습니다.

### 4. 고민과 결론
고민을 하다보니, JPA에서 제공해주는 기능 중에 하나가 1차 캐시를 통한 동일성을 보장해주는 기능을 그냥 사용하면 되지 않을까? 라는 생각이 들었습니다.
그래서 문제가되는 상황이 있을까 생각해보았습니다.
###### 비영속
이 상태에서 엔티티 객체는 데이터베이스에 존재하지 않기 때문에 비교하면 안된다고 생각합니다.

###### 영속
영속 상태에서는 1차 캐시를 통해 동일성이 보장됨

###### 준영속
준영속 상태가 된 후, 새롭게 조회한 객체와 동일성 비교를 하면 보장되지 않음
- 다만, 준영속 상태가 된 객체를 사용할 일이 존재할까?
	- entity Manager가 종료됨, 닫힘 -> service 내에서 그럴 일 없음
	- 직접 detach -> 그럴일 없음

###### 삭제
삭제되면 비교하면 안된다고 생각합니다.

고민해보니 큰 문제가 될 것이 없고 제공해주는 기능을 사용해보기로 했습니다.!!

changuii replied:
#### 추가

오늘 오전에, 크루를 설명해주면서 lombok으로 다시 테스트를 해보니 성공하고 있어서 왜 그런지 의문이 들어 실제 lombok을 delombok해서 코드를 직접 들여다보았습니다.

```java
    @Override
    public boolean equals(final Object o) {
        if (o == this) return true;
        if (!(o instanceof Orders)) return false;
        final Orders other = (Orders) o;
        if (!other.canEqual((Object) this)) return false;
        final Object this$id = this.getId();
        final Object other$id = other.getId();
        if (this$id == null ? other$id != null : !this$id.equals(other$id)) return false;
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof Orders;
    }

    @Override
    public int hashCode() {
        final int PRIME = 59;
        int result = 1;
        final Object $id = this.getId();
        result = result * PRIME + ($id == null ? 43 : $id.hashCode());
        return result;
    }
```

제가 앞에서 말하고 있던, getClass()가 아니라 instanceof를 사용하고 있어 문제가 없었고, getId()를 사용하고 있어 필드에 있는 id를 사용하고 있지도 않아 문제가 발생하지 않고 있었습니다.

- 혹시 제가 매트의 의도를 잘못 파악한 것일까요?

hyeonic commented on src/main/java/roomescape/reservation/domain/ReservationDate.java:
```diff
@@ -0,0 +1,51 @@
+        return CACHE.computeIfAbsent(date, ReservationDate::new);
+    }
+
+    public static ReservationDate fromQuery(final LocalDate date){
+        return CACHE.computeIfAbsent(date, ReservationDate::new);
+    }
```
```suggestion
    public static ReservationDate fromQuery(final LocalDate date) {
        return CACHE.computeIfAbsent(date, ReservationDate::new);
    }
```

컨벤션 확인해주세요!

changuii replied:
신경쓰겠습니다. 감사합니다 :)

hyeonic commented on src/main/java/roomescape/reservation/domain/ReservationDate.java:
```diff
@@ -0,0 +1,51 @@
+import lombok.RequiredArgsConstructor;
+import roomescape.exception.custom.reason.reservation.ReservationPastDateException;
+
+@Embeddable
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+public class ReservationDate {
+
+    private static final Map<LocalDate, ReservationDate> CACHE = new HashMap<>();
```
캐시가 필요한 이유가 뭔가요?? 추가로 HashMap 사용은 여러 쓰레드가 동시 접근할 때 어떠한 문제가 발생할 수 있을까요?

changuii replied:
VO는 불변 객체로 만들었기 때문에 캐시를 통해 계속 만들 필요 없이 사용했던 ReservationDate는 다시 사용해도 괜찮을 것이라 판단하였습니다.
매트가 말씀하신 대로 HashMap은 여러 스레드가 동시에 접근할 때 동시성 문제가 생길 수 있어 동시성 문제를 고려하여 만든 ConcurrentHashMap으로 구현체를 변경해주었습니다.!!

hyeonic replied:
넵 좋습니다. 추후 기회가 된다면 ConcurrentHashMap이 어떠한 방법으로 동시성 문제를 개선하였는지 알아보면 좋을 것 같아요 ㅎㅎ

changuii replied:
concurrentHashMap은 해시 값을 기준으로 컬렉션을 여러 버킷으로 분할하여 처리합니다.
때문에, 작은 단위로 동시성 락을 걸어주어 성능을 향상시켰습니다.
- 락의 범위가 작을수록 여러 스레드가 동시에 처리할 수 있는 공간이 많아지기 때문에 기다리는 스레드가 줄어들고 결과적으로 성능이 높아짐

또한, volatile 키워드를 통해 메모리 가시성을 보장하기 때문에 여러 스레드들이 다른 코어에서 실행되더라도 모두 같은 최신 데이터를 바라보고 있어 조회를 할때는 락을 사용하지 않고 조회합니다.


락을 사용할 때도, synchronized만 사용하는 것이 아니라 CPU 수준의 CAS 연산을 통해서 동시성 문제를 해결하여 스레드가 기다리는 것 없이 연산을 바로 진행합니다.
- CAS : 현재 값을 조회한 후, 값을 수정하기 전에 메모리의 현재 값이 조회한 값이 동일하다면 값을 수정합니다. 이때 각 단계가 CPU 수준에서 원자적(더이상 쪼개질 수 없음)이므로, 동시성이 해결됩니다.! 


#### CPU 코어가 여러개인 경우, 병렬 처리가 되어 버리면 여기서 또 동시성 문제가 발생할 수 있을 것 같은데 CAS가 어떻게 가능할까?
위의 내용이 궁금해서 찾아보았습니다. 자바에서 제공하는 네이티브 메서드를 통해서 CAS 연산을 수행하면 CPU 수준으로 내려가 연산을 진행하는데 CPU가 이 연산들을 원자적으로 수행하기 위해 여러 코어에서 병렬로 실행되지 못하게 락을 걸거나, 하나의 코어만 이 연산을 성공할 수 있도록 만들어 준다고 합니다.!

hyeonic replied:
크으.. 아주 꼼꼼하게 학습하셨네요 👍

 concurrent 관련 자료구조에서는 대부분 CAS를 활용하고 있어서 관련 개념 잘 습득해두면 이후 비슷한 키워드도 쉽게 이해하고 넘어갈 수 있을 것 같아요 ㅎㅎ

hyeonic commented on src/test/java/roomescape/common/PaymentManagerTest.java:
```diff
@@ -0,0 +1,114 @@
+import roomescape.common.dto.PaymentRequest;
+import roomescape.exception.custom.reason.payment.PaymentConfirmException;
+
+@RestClientTest(PaymentManager.class)
+class PaymentManagerTest {
+
+    private final PaymentManager paymentManager;
+    private final MockRestServiceServer server;
+
+    @Autowired
+    public PaymentManagerTest(
+            final PaymentManager paymentManager,
+            final MockRestServiceServer server
+    ) {
+        this.paymentManager = paymentManager;
+        this.server = server;
+    }
```
이렇게 외부 API를 사용할 때 mocking시 어떠한 이점을 얻을 수 있을까요??

changuii replied:
###### 1. 토스 api를 포함하지 않고 내가 작성한 부분만 테스트 가능
외부 API를 사용할 때 우리가 제어할 수 없는 부분을 mocking하여 특정 행동에 대한 제가 작성한 부분을 테스트할 수 있었습니다.
- 예를 들어, toss에서 400 BadRequest를 응답으로 전달했을 때 PaymentManager는 이렇게 동작해야 한다를 테스트하였습니다.

###### 2. 테스트가 외부에 영향을 받지 않음
또한, 외부 API에 테스트가 영향을 받지 않습니다. 실제 toss api를 호출하여 테스트를 작성할 경우 toss 서버의 상태 또는 변화에 따라 테스트가 실패할 수 있지만, mocking을 통해 이를 제어했기 때문에 이런 문제들을 고려하지 않아도 됩니다.

###### 3. 테스트의 속도가 빠름
외부 api를 테스트에서 호출 시 네트워크를 통해 응답을 받아오는 시간까지 기다려야 하지만, mocking하여 프로세스 외부로 패킷이 나갔다가 돌아오는 시간을 포함하지 않기 때문에 테스트의 속도가 빠릅니다.

hyeonic reviewed (changes requested):
안녕하세요 부기! 레벨 2 마지막 미션 방탈출 결제 미션 리뷰를 맡은 매트입니다.

간단히 코멘트 남겼으니 확인해주시고 추가적으로 궁금한 부분 있다면 언제든 편하게 DM 남겨주세요~

changuii committed:
fix: themeId에 timeId를 넣던 버그 수정

changuii committed:
fix: @RequiredArgsConstructor 제거

changuii committed:
fix: 동시성 문제가 발생할 수 있는, HashMap을 ConcurrentHashMap으로 변경

changuii committed:
refactor: TimeManager를 통해 시간을 받아오도록 변경

changuii committed:
style: 코드 스타일 수정 (컨벤션 준수 및 간격, 정렬 개선)

- Java 코드의 공백, 괄호, 줄 바꿈 등 컨벤션 준수를 위한 스타일 수정
- 가독성을 향상시키기 위해 메서드와 쿼리 정의에서 공백 및 정렬 조정

changuii commented:
미션 커밋입니다.!!! 매트 리뷰 감사합니다 :)
https://github.com/woowacourse/spring-roomescape-payment/pull/190/files/a743ca8ff23bfe372ed6ef2e7f13e294ae897237..fc467d8c6515edd81afef28810682e535d1b1431

hyeonic commented on src/main/java/roomescape/common/PaymentManager.java:
```diff
@@ -0,0 +1,55 @@
+// toss pg사 이용
+@Component
+public class PaymentManager {
+
+    private final ObjectMapper objectMapper;
+    private final RestClient restClient;
+
+    @Autowired
+    public PaymentManager(
+            final RestClient.Builder restClientBuilder,
+            final ObjectMapper objectMapper,
+            @Value("${secret.toss-payment-key}") final String authorizationKey
+    ) {
+        this.objectMapper = objectMapper;
+
+        restClient = restClientBuilder
+                .baseUrl("https://api.tosspayments.com")
+                .defaultHeader("Authorization", authorizationKey)
+                .build();
+    }
+
+    public void confirmPayment(final PaymentRequest paymentRequest) {
+        try {
+            restClient.post()
+                    .uri(URI.create("/v1/payments/confirm"))
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .body(paymentRequest)
+                    .retrieve()
+                    .onStatus(HttpStatusCode::isError, ((request, response) -> {
+                        final String responseBody = new String(response.getBody().readAllBytes());
+                        final PaymentError error = objectMapper.readValue(responseBody, PaymentError.class);
+                        throw new PaymentConfirmException(error);
+                    }))
+                    .toBodilessEntity();
+        } catch (final ResourceAccessException e) {
+            final PaymentError error = PaymentError.connectFail();
+            throw new PaymentConfirmException(error);
+        }
+    }
+}
```
현재는 toss에 너무 강하게 의존하는 것 처럼 보이네요. 추후 다른 결제 회사를 고려하거나 추가될 경우 유연하게 대응할 수 있을까요??

changuii replied:
제가 생각했던 부분은 다음과 같습니다! ! 

### 새로운 결제 회사가 추가되는 경우
먼저 새로운 결제 회사가 추가되는 경우는 딱히 고려를 하지 않았습니다. 왜냐하면, 현재 토스 페이먼트 PG를 사용하고 있는데, Payment Gateway가 여러 결제 회사와 연동을 하기 어렵기 때문에 이것을 해결해주기 위해 하나의 Gateway로 여러 결제 회사를 제공해줄 수 있도록 만들었다고 생각했습니다. 그래서 이런 Gateway를 사용하고 있기 때문에 굳이 다른 회사를 추가하는 것은 고려하지 않아도 된다고 생각했습니다. 혹시 Gateway를 여러 가지를 사용하는 경우도 가까운 미래에 해당하여 고려해야하는 부분인지 궁금합니다!

### 새로운 결제 회사로 변경을 고려하는 경우
이 경우도 PaymentManager, request DTO를 변경해주면 service는 큰 변화가 없을 것이라 판단하였습니다.
만약, 결제 방식 혹은 프로세스가 변경되어 service의 변경이 일어나더라도, 그건 추상화 계층을 두고 변경을 막으려고 하더라도 결국 service 프로세스의 변경이 필요한 일이기 때문에 어쩔 수 없는 변경이라 판단하였습니다.

혹시 다른 경우가 있을까요?? 아니면 제가 미처 생각하지 못한 부분이 존재하나요?? 매트의 생각이 궁금합니다.

hyeonic replied:
저는 PaymentManager에 toss라는 구체적인 정보로 진행하게 되면 추후 다른 결제 회사로 변경되었을 때 PaymentManager의 직접적인 수정이 결국 서비스까지 전파되지 않을까 우려되었어요 ㅎㅎ

우리가 개발하고 있는 방탈출 결제 애플리케이션에서 집중하고자 하는 도메인 영역은 `결제`이지 `toss 결제`가 되어서는 안된다고 생각합니다. 언제든 결제 수단은 추가될 수 있을 뿐만 아니라 우리가 제어할 수 없는 영역이 모종의 이유로 수정되었을 때 결국 저희가 지키고자 하는 도메인 영역까지 수정되는 것을 최소화하는 방법을 지속적으로 고민해보아야 한다고 생각했어요. 결제를 위해서 필요한 정보를 toss에 의존하기 보단 좀 더 폭 넓게 방탈출 결제 시 필요한 정보를 추상화하여 인터페이스를 구성하게 되면 이후 해당 정보를 기반으로 구현체를 구성하여 여러 변경에 대응할 수 있지 않을까 합니다.

changuii replied:
매트에 매트와 허들을 통해 이해했습니다.
우리가 인터페이스를 통해 구현체들에게 "이 인터페이스를 따라서 만들어!"에 대해서 강제해줄 수 있고 이를 통해 구현체들의 변경이 도메인에 영향을 끼치지 않도록 해줄 수 있습니다.
PaymentManager 인터페이스를 두고, 구현체인 TossPaymentManager를 사용한다면, 실제로 TossPaymentManager가 변경이 필요할 때 정말 필요한 변경이 아니라면 해당 구현체 내부에서 원래의 인터페이스를 따르도록 강제시킬 수 있다고 이해를 했습니다.

즉, 인터페이스를 둠으로써 의존 관계가 역전되어 구현체가 우리 도메인의 인터페이스를 따르도록 하고, 우리 도메인 계층의 ReservationService는 도메인 내부의 인터페이스에 의존하게 되어 결론적으로, "어쩔 수 없는 변경"이라면 그대로 변경이 생기겠지만 그렇지 않은 경우에 우리 도메인의 변경이 아니라 구현체의 변경만으로 끝날 수 있도록 강제성이 생겨 우리 도메인을 안전하게 지킬 수 있게 됩니다.

이걸 이해하고, 도메인을 어떻게 안전하게 지킬 수 있을까 고민해보고 패키지 구조도 변경했습니다.
도메인 내부에서는 외부를 알 수 없게 만들어, 우리에게 중요한 문제 영역은 변경으로부터 안전하게 지키고, 도메인 외부에서 도메인 내부를 의존하는 구조가 되도록 변경했습니다.
- 결국 DIP 원칙을 통해 변경에 안전한 구조가 된 것 같습니다.!

hyeonic commented on src/main/java/roomescape/exception/custom/reason/ResponseInvalidException.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.http.HttpStatus;
+import roomescape.exception.custom.status.CustomException;
+
+public class ResponseInvalidException extends CustomException {
+    public ResponseInvalidException() {
+        super(HttpStatus.INTERNAL_SERVER_ERROR, "서버에 문제가 발생했습니다. 관리자에게 문의하세요");
+    }
```
부기는 어떠한 상황에서 500대 에러를 고려하시나요??

changuii replied:
논리적으로 발생할 수 없는 상황에는 일단 500대 에러를 응답하도록 고려하였습니다.
예를 들면, response를 만들어서 반환하는 과정에서 null 값이 발생하면 예외가 발생하도록 만들어뒀습니다.

hyeonic replied:
> response를 만들어서 반환하는 과정에서 null 값이 발생하면

요 부분은 개발자가 제어할 수 있는 부분 아닌가요? 응답이 정상적으로 만들어지지 않았을 때의 이유가 좀 더 중요하지 않을까 생각이 들어요! 가령 DB에 접근하지 못하여 응답 객체를 못만드면 이러한 null이 제어할 수 없는 부분 이므로 500일 수 있지만 단순히 조회 결과가 없어서 null인 경우엔 앞단에서 400에러를 내려줄 수도 있지 않을까 생각이 들어서요!

changuii replied:
> 단순히 조회 결과가 없어서 null인 경우엔 앞단에서 400에러를 내려줄 수도 있지 않을까 생각이 들어서요!

매트의 말에 동의합니다. 
다만 현재 데이터베이스에 필드들이 NOT NULL 필드로 구현되어 있어 만약 응답 dto를 만드는 과정에 null이 포함된다면 "database 조회 후"와 "응답 dto 생성" 과정 사이에 entity의 필드가 null로 변경되는 실수가 발생했다는 의미라고 생각합니다.
- 따라서 이 부분이 저는 개발자의 실수라고 판단하여 500을 응답하도록 설계하였습니다.! 

 제가 이전 코멘트에서 말씀드렸던 "response를 만들어서 반환하는 과정에서 null 값이 발생하면"이 상황이 "데이터베이스 조회 후, 응답 dto를 생성하는 과정 사이에 모종의 이유로 null이 입력되는 경우"를 염두하고 말씀드렸습니다.

hyeonic commented on src/main/java/roomescape/reservation/repository/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,99 @@
+import roomescape.reservationtime.domain.ReservationTime;
+import roomescape.theme.domain.Theme;
+
+@Repository
+@RequiredArgsConstructor
+public class ReservationRepositoryImpl implements ReservationRepository {
```
Impl과 같은 표현은 추후 또 다른 구현체가 생겼을 때 이름을 짓기가 모호하더라구요. 또한 단순히 JpaRepository의 메서드를 by pass하는 것으로 보이는데 이러한 구현체가 구성된 이유는 무엇인가요??

changuii replied:
매트의 생각에 저도 동의합니다.! 다만 네이밍을 이렇게 가져간 이유는 해당 인터페이스의 구현체가 추가로 생기기 힘들 것이라는 생각이 들어 일단 이렇게 지었습니다.

처음 이렇게 by pass하는 클래스를 생각한 이유는, JDBC에서 JPA로 변경하는 과정에서 기존의 interface를 바로 JPA Repository로 변경하였습니다. 때문에 많은 fake 객체들을 더이상 사용할 수 없었고, 사용하더라도 불필요한 메서드 구현이 필요했습니다.
또한, 기존에 JDBC를 사용할 때처럼 interface가 모든 메서드들을 정의하고 있지 않다 보니 다소 한눈에 어떤 메서드들이 사용중인지 알기 어려웠습니다.
그리고, JPA를 사용하고 있다가 추후 다른 데이터 접근 기술을 고려할 때 변경하기가 어려워질 것이라 판단하였습니다.
마지막으로, 여러 데이터 접근 기술을 활용하는 경우도 생길 수 있지 않을까? 라는 생각도 들었습니다.

그래서 정리해보면, 테스트 대역 구현의 어려움과 가독성, 데이터 접근 기술 추상화, 여러 데이터 접근 기술 사용의 장점이 by pass 하는 메서드를 만드는 것보다 이점이 크다고 판단하여 이렇게 구조를 만들어주었습니다.

hyeonic replied:
> 해당 인터페이스의 구현체가 추가로 생기기 힘들 것이라는 생각이 들어 일단 이렇게 지었습니다.

그렇다면 굳이 인터페이스라는 개념을 유지할 필요가 있을까요?? 확장의 여지가 없다면 인터페이스를 온전하게 활용하지 못하는 느낌이 들어서요! 

말씀 주신 것 처럼 이전에 JDBC -> JPA로 변경하며 고려하지 못한 메서드 네이밍 등으로 인해 이러한 결과물이 탄생하였다고 생각합니다. 유연하게 바뀌기는 어렵겠지만 이미 스프링에서 `Repository`, `CrudRepository` 등의 인터페이스를 통해 고수준의 추상화 레벨을 제공하고 있어서 저희가 별도로 중간에 Adapter 개념을 두는 것이 괜찮은지는 약간의 의문이 들기도 하는 것 같아요. 요 부분은 제 개인적인 견해가 많이 담겨있어서 무조건적으로 반영하기 보단 참고만 부탁드립니다!

changuii replied:
앞에서도 말했지만, 데이터 접근 계층의 변경과 테스트 용이성이라는 장점을 가져가기 위해서 이렇게 추상화를 했으나, 다소 큰 의미를 못느낀다고 저도 동의합니다. :)
다만, 직접적으로 JpaRepository를 사용하는게 저는 우려된다고 생각해요.

hyeonic reviewed (changes requested):
안녕하세요 부기! 간단히 코멘트 남겼으니 확인해주시고 추가적으로 궁금한 부분이 있다면 언제든 편하게 DM 남겨주세요~

changuii committed:
refactor: 불필요한 EqualsAndHashCode 및 ToString 어노테이션 제거

- 주요 도메인 클래스(ReservationTime, Reservation, Theme, Member 등)에서 필요하지 않은 @EqualsAndHashCode, @ToString 어노테이션 제거하여 코드 간소화
- ReservationTimeService에 @Transactional 추가
- 테스트 코드에서 ArgumentMatchers.argThat을 활용하여 삭제 메서드의 인자 검증 추가

hyeonic commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -0,0 +1,83 @@
+import roomescape.reservationtime.domain.ReservationTime;
+import roomescape.theme.domain.Theme;
+
+@Entity
+@Getter
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class Reservation {
```
equals & hashCode 관련 코멘트 아래서 좀 더 진행해볼게요!

 * [코멘트 1](https://github.com/woowacourse/spring-roomescape-payment/pull/190#discussion_r2119221722)
 * [코멘트 2](https://github.com/woowacourse/spring-roomescape-payment/pull/190#discussion_r2120026034)

다양한 실험 사례 기반으로 잘 정리해주셨는데요! 먼저 제가 처음에 질문 드린 의도는 lombok 통해서 생성된 코드로 실제 객체와 프록시 객체 간의 비교도 문제가 없는지 궁금하여 질문드렸어요.

또한 `@EqualsAndHashCode(onlyExplicitlyIncluded = true)`을 제가 delombok하여 코드를 본 것이 아니라 `instanceof`를 통해 비교를 하는 부분 또한 인지 하지 못하였습니다. 말씀 주신 것 처럼 instanceof를 통해 프록시 객체 여부를 체크하고 getId()를 통해 영속된 id를 조회할 경우 문제가 없을 것으로 판단합니다! 

관련 로직 제거하지 않아도 괜찮을 것 같은데 어떻게 생각하시나요??

changuii replied:
동의합니다!! 

다만, 매트에게 리뷰 요청을 보낸 후, [코멘트 2](https://github.com/woowacourse/spring-roomescape-payment/pull/190#discussion_r2120026034)를 인지하여, 수정하지는 못했습니다.

추가로, 위에 제가 코멘트에 작성하지는 않았으나, 영속화 되기 이전에 Set 컬렉션을 사용하거나 비영속 상태의 엔티티를 비교하는 경우에 문제가 발생할 수 있다는 것도 인지하였으나 영속 상태가 아닌 엔티티를 사용하는 것 자체가 문제라고 판단하여 이는 고려하지 않았습니다.

다시 `@EqualsAndHashCode` 추가해두었습니다. :)

hyeonic reviewed (changes requested):
안녕하세요 부기~ 마지막 미션이라 외부 API 호출 이외에도 다방면으로 질문드렸음에도 대부분 잘 구현해주시고 답변도 잘 남겨주셨는데요! 좀 더 확인하고 넘어가고 싶은 부분이 있어서 추가 코멘트 남겼습니다 ㅎㅎ 

아마 관련 논의만 마무리하면 이후 단계 진행해도 좋을 것 같아서 간단히 확인만 부탁드립니다. 제 추가 질문이 모호하거나 좀 더 궁금한 부분이 있다면 언제든 DM 남겨주세요!

changuii committed:
refactor: 엔티티에 `@EqualsAndHashCode` 추가

changuii committed:
refactor(payment): Payment 관련 클래스 및 테스트 패키지 구조 변경

Payment 관련 클래스와 테스트를 common 패키지에서 payment 패키지로 이동하여 모듈화 및 패키지 구조 정리.

changuii committed:
refactor: AdminViewController와 UserViewController의 패키지 경로를 roomescape.view로 변경

changuii committed:
refactor: TimeManager 인터페이스 분리 및 패키지 구조 개선

- Theme과 Reservation에서 사용되는 TimeManager를 ThemeTimeManager와 ReservationTimeManager로 분리
- Payment 관련 클래스들을 dto 패키지로 이동
- 공통 유틸 TimeManager를 infrastructure 패키지로 이동 및 인터페이스 구현

changuii committed:
refactor: 도메인 및 인프라 계층 패키지 구조 분리

- domain, infrastructure, controller 간의 책임 분리를 위해 패키지 재구성
- Auth, Reservation, ReservationTime, Member, Theme 관련 클래스 및 인터페이스를 역할에 따라 이동
- 인터페이스 및 구현체 명 명시적으로 리팩토링

hyeonic reviewed (approved):
안녕하세요 부기! 대부분 잘 답변 남겨주시고 반영해주셔서 이후 단계 진행해도 좋을 것 같습니다. 고생하셨습니다 😃
